### PR TITLE
Update repository size on cron gc task

### LINKED ISSUE #14862: Repo size on web isn't updated after Auto GC
- Gitea version (or commit ref): master (commit b2986ab8b2671c2779aa4a8b72cb7947c62402a4)
- Git version: 2.30.1
- Operating system: Debian 10
  
  
  
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No (not relevant)
- Log gist: (not relevant)

## Description

When auto gc occurs, the actual repo size is shrinked, but the repo size on web interface isn't updated accordingly.

For example, today I pushed recent commits on `git.git` (Git SCM repo) to my local Gitea instance, then auto gc was invoked. 
After auto gc, the repo size was about 132 MiB, but on the web interface the size was still 211 MiB (before gc)

## Screenshots

on web interface
*on web interface, repo size: 211 MiB*
git count-objects -v on the instance
*git count-objects -v when accessing the interface, repo size: 132 MiB*

### PR DESCRIPTION
**Pull request by zeripath**
git gc cron could change the size of the repository therefore we should update the
size of the repo stored in our database.

Also significantly improve the efficiency of counting lfs associated with the
repository

Fix #14862

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by lunny — APPROVED** [review]

File: modules/repository/check.go [hunk 1]
```diff
@@ -91,6 +91,17 @@ func GitGcRepos(ctx context.Context, timeout time.Duration, args ...string) erro
...
 			}
+
+			// Now update the size of the repository
+			if err := repo.UpdateSize(models.DefaultDBContext()); err != nil {
+				log.Error("Updating size as part of garbage collection failed for %v. Stdout: %s\nError: %v", repo, stdout, err)
+				desc := fmt.Sprintf("Updating size as part of  garbage collection failed for %s. Stdout: %s\nError: %v", repo.RepoPath(), stdout, err)
+				if err = models.CreateRepositoryNotice(desc); err != nil {
+					log.Error("CreateRepositoryNotice: %v", err)
+				}
+				return fmt.Errorf("Updating size as part of  garbage collection failed in repo: %s: Error: %v", repo.FullName(), err)
```
Comment (lafriks) [inline modules/repository/check.go]:
```suggestion
				return fmt.Errorf("Updating size as part of garbage collection failed in repo: %s: Error: %v", repo.FullName(), err)
```

File: modules/repository/check.go [hunk 2]
```diff
@@ -91,6 +91,17 @@ func GitGcRepos(ctx context.Context, timeout time.Duration, args ...string) erro
 				}
 				return fmt.Errorf("Repository garbage collection failed in repo: %s: Error: %v", repo.FullName(), err)
 			}
+
+			// Now update the size of the repository
+			if err := repo.UpdateSize(models.DefaultDBContext()); err != nil {
+				log.Error("Updating size as part of garbage collection failed for %v. Stdout: %s\nError: %v", repo, stdout, err)
+				desc := fmt.Sprintf("Updating size as part of  garbage collection failed for %s. Stdout: %s\nError: %v", repo.RepoPath(), stdout, err)
```
Comment (lafriks) [inline modules/repository/check.go]:
```suggestion
				desc := fmt.Sprintf("Updating size as part of garbage collection failed for %s. Stdout: %s\nError: %v", repo.RepoPath(), stdout, err)
```

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
🚀

**Comment by silverwind** [comment]
Related regression: https://github.com/go-gitea/gitea/issues/15634

**Pull request merged**